presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java

```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
+        }
+
+        @Override
+        public Optional<Signature> bind(List<? extends Type> actualArgumentTypes)
+        {
+            ImmutableList.Builder<TypeSignature> boundTypesSignatures = ImmutableList.builder();
+
+            long lengthAccumulator = 0;
+            for (Type type : actualArgumentTypes) {
+                if (!isTypeValid(type)) {
+                    return Optional.empty();
+                }
+
+                int paramLength = varcharLength(type);
+                lengthAccumulator = Ints.saturatedCast(lengthAccumulator + paramLength);
+                boundTypesSignatures.add(VarcharType.createVarcharType(paramLength).getTypeSignature());
+            }
+
+            return Optional.of(
+                    new Signature(
+                            declaredSignature.getName(),
+                            declaredSignature.getKind(),
+                            VarcharType.createVarcharType(Ints.saturatedCast(lengthAccumulator)).getTypeSignature(),
+                            boundTypesSignatures.build()));
+        }
+
+        @Override
+        public Optional<BoundVariables> bindVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)
+        {
+            Optional<Signature> signature = bind(actualArgumentTypes);
+            if (!signature.isPresent()) {
+                return Optional.empty();
+            }
+
+            if (!isTypeValid(actualReturnType)) {
+                return Optional.empty();
+            }
+
+            int returnLength = varcharLength(actualReturnType);
+            int calculatedReturnLength = Ints.saturatedCast(getOnlyElement(signature.get().getReturnType().getParameters().iterator()).getLongLiteral());
+            if (returnLength < calculatedReturnLength) {
+                return Optional.empty();
+            }
+
+            ImmutableMap.Builder<String, Long> longBoundBuilder = ImmutableMap.builder();
+            longBoundBuilder.put("r", new Long(returnLength));
+            int i = 0;
+            for (TypeSignature argSignature : signature.get().getArgumentTypes()) {
+                longBoundBuilder.put("x" + Integer.toString(i++), getOnlyElement(argSignature.getParameters().iterator()).getLongLiteral());
+            }
+
+            // this is not required, but is it passed to make it clear how things do bind
+            return Optional.of(new BoundVariables(ImmutableMap.of(), longBoundBuilder.build()));
+        }
+
+        private boolean isTypeValid(Type type)
+        {
+            if (allowCoercion && typeManager.coerceTypeBase(type, StandardTypes.VARCHAR).isPresent()) {
+                return true;
+            }
+
+            if (type.getTypeSignature().getBase() != StandardTypes.VARCHAR) {
+                return false;
+            }
+
+            TypeSignatureParameter param = getOnlyElement(type.getTypeSignature().getParameters().iterator());
+            checkArgument(param.getKind() == ParameterKind.LONG);
```

Comment:
isn't this check redundant? Parameters should be verified when `VarcharType` is created.